### PR TITLE
Iox #25 template config via properties

### PR DESCRIPTION
**Pull request by marthtz**
Introduce template configuration via property structs for chunk distributor data and chunk queue data.

This PR depends on #202 to be merged first as it includes some commits of that one.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+!building_blocks/
```
Comment (mossmaurice) [inline .gitignore:1]: Why this change?

Comment (marthtz) [reply .gitignore:1]: Our `.gitignore` file includes `build*/`. That would also ignore all `building_blocks` folders. I added this line so `building_blocks` are exempt from ignoring.

Comment (ithier) [reply .gitignore:1]: Woops >_>

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp [hunk 2]
```diff
@@ -26,24 +26,24 @@ namespace iox
...
 {
+template <typename ChunkQueueDataProperties>
 struct ChunkQueueData
 {
+    using ChunkQueueDataProperties_t = ChunkQueueDataProperties;
+
     explicit ChunkQueueData(cxx::VariantQueueTypes queueType) noexcept;
 
-    static constexpr uint32_t MAX_CAPACITY = MAX_RECEIVER_QUEUE_CAPACITY;
+    static constexpr uint32_t MAX_CAPACITY = ChunkQueueDataProperties_t::MAX_QUEUES;
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp:36]: Maybe this is nor more needed if we could now use ChunkQueueData<..>::ChunkQueueDataProperties_t ::MAX_QUEUES? 

On the other hand, this one more line will make others more readable...

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl [hunk 3]
```diff
@@ -11,19 +11,20 @@
...
 {
 namespace popo
 {
-ChunkReceiverData::ChunkReceiverData(const cxx::VariantQueueTypes queueType,
-                                     const mepoo::MemoryInfo& memoryInfo) noexcept
-    : ChunkQueueData(queueType)
-    , m_memoryInfo(memoryInfo)
+template <typename ChunkQueueProperties>
+inline ChunkQueueData<ChunkQueueProperties>::ChunkQueueData(cxx::VariantQueueTypes queueType) noexcept
+    : m_queue(queueType)
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl:23]: You removed the memoryInfo. I guess it was not used yet but introduced by @elfenpiff for later usage, I guess

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl [hunk 4]
```diff
@@ -0,0 +1,175 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl:1]: I guess no changes here, only a rename from cpp to inl...

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl [hunk 5]
```diff
@@ -59,14 +68,12 @@ cxx::expected<ChunkQueueError> ChunkQueuePusher::push(mepoo::SharedChunk chunk)
             auto returnedChunk = mepoo::SharedChunk(chunkManagement);
         }
 
-
-        if (getMembers()->m_semaphoreAttached.load(std::memory_order_acquire) && getMembers()->m_semaphore)
+        if (getMembers()->m_conditionVariableAttached.load(std::memory_order_acquire)
+            && getMembers()->m_conditionVariableDataPtr)
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_pusher.inl:72]: Maybe we should concentrate on merging #202 as your PR also includes this changes...

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp [hunk 6]
```diff
@@ -36,10 +36,11 @@ enum class ChunkReceiveError
 /// The
 /// ChunkRceiver holds the ownership of the SharedChunks and does a bookkeeping which chunks are currently passed to the
 /// user side.
-class ChunkReceiver : public ChunkQueuePopper
+template <typename ChunkQueueDataType>
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp:39]: shouldn't this be ChunkQueueType here to be symmetric?

ChunkReceiverData has ChunkQueueDataType as template and then ChunkReceiver has the ChunkQueueType

OK, then you would inherit from something like ChunkQueuePopper<ChunkQueueType::ChunkQueueData_t>

But I feel this would be more one layer uses the one below

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp [hunk 7]
```diff
@@ -35,11 +36,14 @@ struct ChunkReceiverData : public ChunkQueueData
     /// we use one more than MAX_CHUNKS_HELD_PER_RECEIVER for being able to provide one new chunk
     /// to the user if they already have the allowed MAX_CHUNKS_HELD_PER_RECEIVER. But then the user
     /// has to return one to not brake the contract. This is aligned with AUTOSAR Adaptive ara::com
-    static constexpr uint32_t MAX_CHUNKS_IN_USE = MAX_CHUNKS_HELD_PER_RECEIVER + 1u;
+    static constexpr uint32_t MAX_CHUNKS_IN_USE =
+        ChunkQueueDataType::ChunkQueueDataProperties_t::MAX_CHUNKS_PER_RECEIVER + 1u;
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp:40]: I thing I don't like from design point of view is that your ChunkQueueDataProperties include the MAX_CHUNKS_PER_RECEIVER. This is not used by the ChunkQueue itself but later in the class hierarchy by ChunkReceiver. Maybe this should be a separate config for ChunkReceiverData, or maybe add this single template parameter as having a config struct for one alone is a bit of overhead

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp [hunk 8]
```diff
@@ -36,7 +36,7 @@ struct ChunkSenderData : public ChunkDistributorDataType
 
     const relative_ptr<mepoo::MemoryManager> m_memoryMgr;
     mepoo::MemoryInfo m_memoryInfo;
-    UsedChunkList<MAX_CHUNKS_ALLOCATE_PER_SENDER> m_chunksInUse;
+    UsedChunkList<ChunkDistributorDataType::ChunkDistributorDataProperties_t::MAX_CHUNKS_PER_SENDER> m_chunksInUse;
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp:39]: same same but different

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp [hunk 9]
```diff
@@ -91,7 +91,7 @@ class SubscriberPortUser
 
     MemberType_t* m_subscriberPortDataPtr;
 
-    ChunkReceiver m_chunkReceiver;
+    ChunkReceiver<SubscriberPortData::ChunkQueueData_t> m_chunkReceiver;
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp:94]: symmetric to the sender side would be then
    `using ChunkQueue_t = ChunkQueue<SubscriberPortData::ChunkQueueData_t>;`
    `ChunkReceiver<ChunkQueue_t > m_chunkReceiver;`

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 10]
```diff
@@ -123,6 +122,21 @@ enum class SubscribeState : uint32_t
     WAIT_FOR_OFFER
 };
 
+// Default properties of ChunkDistributorData
+struct DefaultChunkDistributorConfig
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:126]: see also below, your config for ChunkDistributor includes MAX_CHUNKS_PER_SENDER which is used in ChunkSender then. Same for MAX_CHUNKS_PER_RECEIVER  in the config for ChunkQueue. Separation of concerns here

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl (hunk 3 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl:23]: This was part of @mossmaurice commit in #202.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl (hunk 4 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl:1]: Corrrect. For some reason it's not identified as such for this file.

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 11]
```diff
@@ -123,6 +122,21 @@ enum class SubscribeState : uint32_t
...
+{
+    static constexpr uint32_t MAX_QUEUES = MAX_SUBSCRIBERS_PER_PUBLISHER;
+    static constexpr uint32_t MAX_CHUNKS_PER_SENDER = MAX_CHUNKS_ALLOCATE_PER_SENDER;
+    static constexpr uint64_t MAX_HISTORY_CAPACITY = MAX_HISTORY_CAPACITY_OF_CHUNK_DISTRIBUTOR;
+};
+
+// Default properties of ChunkQueueData
+struct DefaultChunkQueueConfig
+{
+    static constexpr uint32_t MAX_QUEUES = MAX_RECEIVER_QUEUE_CAPACITY;
```
[the remark this answers is not in the document]
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp]: Done.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp (hunk 2 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp]: That's what I had in my mind too. Either way is fine for me. What about the others?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 10 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:126]: Yeah, got it. Put it now as separate param in chunk sender/receiver data. Have a look.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp (hunk 7 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp]: Yeah, got it. Put it now as separate param in chunk sender/receiver data. Have a look.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp (hunk 8 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp]: Yeah, got it. Put it now as separate param in chunk sender/receiver data. Have a look.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp (hunk 9 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp:94]: There's no `ChunkQueue` right now. There's only the queue popper. Should we rework the structure in EA?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp (hunk 6 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp:39]: There's no `ChunkQueue` right now. There's only the queue popper. Should we rework the structure in EA?

File: .gitignore (hunk 1 above)
Comment (marthtz) [reply .gitignore:1]: You're kind of right. Modifying it to `/build*/` should limit it to folders from root folder only.
I'll fix that.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp (hunk 6 above)
Comment (budrus) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp:39]: Right, there is no spoon. It must be ChunkQueuePopper

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp (hunk 9 above)
Comment (budrus) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp:94]: Also here then using `ChunkQueuePopper_t = ChunkQueuePopper<SubscriberPortData::ChunkQueueData_t>;` 

Sorry for confusing

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp (hunk 2 above)
Comment (budrus) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.hpp]: Fine for me as it is

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl (hunk 3 above)
Comment (budrus) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl:23]: This was already merged... Please check if you roll this back. Now it is no more consistent to the other building blocks
FYI @mossmaurice @elfenpiff

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp (hunk 6 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp]: Done

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp (hunk 9 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/ports/subscriber_port_user.hpp]: Done

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl (hunk 3 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_data.inl:23]: Looks like a git mixup:  ...o/building_blocks/chunk_receiver_data.cpp → ...popo/building_blocks/chunk_queue_data.inl 
Chunk queue data never had `memoryInfo` and chunk receiver data still has.

File: iceoryx_posh/include/iceoryx_posh/internal/capro/capro_message.hpp [hunk 12]
```diff
@@ -81,7 +80,7 @@ class CaproMessage
     ServiceDescription m_serviceDescription;
     /// @brief Null-Pointer for request-port with no specific type
     popo::ReceiverPortData* m_requestPort{nullptr};
-    popo::ChunkQueueData* m_chunkQueueData{nullptr};
+    void* m_chunkQueueData{nullptr};
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/internal/capro/capro_message.hpp:83]: Not a big fan of this, but I don't have a better idea either :-/

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp [hunk 13]
```diff
@@ -25,7 +25,8 @@ namespace iox
 {
 namespace popo
 {
-struct ChunkReceiverData : public ChunkQueueData
+template <uint32_t MaxChunksPerReceiver, typename ChunkQueueDataType>
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp:28]: The parameter MaxChunksPerReceiver controls how many chunks the user is allowed to hold. I.e. how many memory can be blocked by chunks that are in processing. For me this is not reflected in this name. Better would be something like MaxChunksAllowedToBeHeld

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp [hunk 14]
```diff
@@ -27,7 +27,7 @@ namespace iox
 {
 namespace popo
 {
-template <typename ChunkDistributorDataType>
+template <uint32_t MaxChunksPerSender, typename ChunkDistributorDataType>
```
Comment (budrus) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp:30]: And here maybe something like MaxChunksAllocatedInParallel

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp (hunk 13 above)
Comment (budrus) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp:28]: Or compared to the sending side something like MaxChunksProcessedInParallel

Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_receiver_data.hpp:28]: What about "MaxChunksHeldSimultaneously"?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp (hunk 14 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender_data.hpp:30]: What about "MaxChunksAllocatedSimultaneously"
